webapp/channels/src/components/convert_gm_to_channel_modal/convert_gm_to_channel_modal.tsx

```diff
@@ -0,0 +1,220 @@
+    teammateNameDisplaySetting: string;
+    channelsCategoryId: string | undefined;
+    currentUserId: string;
+}
+
+const ConvertGmToChannelModal = (props: Props) => {
+    const intl = useIntl();
+    const {formatMessage} = intl;
+
+    const [channelName, setChannelName] = useState<string>('');
+    const handleChannelNameChange = useCallback((newName: string) => {
+        setChannelName(newName);
+    }, []);
+
+    const [channelURL, setChannelChannelURL] = useState<string>('');
+    const handleChannelURLChange = useCallback((newURL: string) => {
+        setChannelChannelURL(newURL);
+    }, []);
+
+    const [channelMemberNames, setChannelMemberNames] = useState<string[]>([]);
+
+    useEffect(() => {
+        const validProfilesInChannel = props.profilesInChannel.
+            filter((user) => user.id !== props.currentUserId && user.delete_at === 0).
+            map((user) => displayUsername(user, props.teammateNameDisplaySetting));
+
+        setChannelMemberNames(validProfilesInChannel);
+    }, [props.profilesInChannel]);
+
+    const [commonTeamsById, setCommonTeamsById] = useState<{[id: string]: Team}>({});
+    const [commonTeamsFetched, setCommonTeamsFetched] = useState<boolean>(false);
+    const [loadingAnimationTimeout, setLoadingAnimationTimeout] = useState<boolean>(false);
+    const [selectedTeamId, setSelectedTeamId] = useState<string>();
+    const [nameError, setNameError] = useState<boolean>(false);
+    const [conversionError, setConversionError] = useState<string>();
+
+    const handleTeamChange = useCallback((teamId: string) => {
+        setSelectedTeamId(teamId);
+    }, []);
+
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        const work = async () => {
+            const response = await dispatch(getGroupMessageMembersCommonTeams(props.channel.id)) as ActionResult<Team[], ServerError>;
+            if (response.error || !response.data) {
+                return;
+            }
+            const teams = response.data;
+
+            const teamsById: {[id: string]: Team} = {};
+            teams.forEach((team) => {
+                teamsById[team.id] = team;
+            });
+            setCommonTeamsById(teamsById);
+            setCommonTeamsFetched(true);
+
+            // if there is only common team, selected it.
+            if (teams.length === 1) {
+                setSelectedTeamId(teams[0].id);
+            }
+        };
+
+        work();
+        setTimeout(() => setLoadingAnimationTimeout(true), 1200);
+    }, []);
+
+    const handleConfirm = async () => {
+        if (!selectedTeamId) {
+            return;
+        }
+
+        const {error} = await props.actions.convertGroupMessageToPrivateChannel(props.channel.id, selectedTeamId, channelName.trim(), channelURL.trim());
+
+        if (error) {
+            setConversionError(error.message);
+            return;
+        }
+
+        setConversionError(undefined);
+        if (props.channelsCategoryId) {
```

Comment:
I guess this is to move the channel to the proper category on the new team, right? Or not?

I am curious about the following scenarios:
- Happy path: from Direct Messages category does it move properly to the Channels category of the target team? And does it disappear from all the other teams Direct Messages categories?
- If I had the channel on this team on a custom category, does it stay in such category? (and disappear from whichever category I had it before on other teams)
- If I had the channel on this team in a custom category, and move it to a different team where that category does not exist, where do it ends up? In the Channels category of that team?

And all of this works properly also for mobile both online and offline?

And... does this has to be handled in any special way in the websocket events?

Reply:
That's a very good point. I'd say that it sshould be removed from all teams except for the destination team, and if it was in a custom category, we'd probably keep it there instead of moving it to the Channels category. I'm 0/5 on if we actually keep custom categories though since I can't see that happening at all often, and I could see that being annoyingly tricky to do.

As for websocket events, that's definitely something to double check how it behaves. Since the sidebar is often changed by side effects of other actions, the websocket events around it aren't great. If we detect that a channel changed from a GM to a non-GM, we might want to just reload all the data for the sidebar for simplicity's sake

Reply:
> Happy path: from Direct Messages category does it move properly to the Channels category of the target team? And does it disappear from all the other teams Direct Messages categories?

Yes, correct.

> If I had the channel on this team on a custom category, does it stay in such category? (and disappear from whichever category I had it before on other teams)

Currently it always moves to  the default "Channels" category. But I'm not checking if it belongs to a custom category, and so, not deleting any existing channel-category data. I think I should though.

> If I had the channel on this team in a custom category, and move it to a different team where that category does not exist, where do it ends up? In the Channels category of that team?

it will always move to custom channels category.

i'm moving the category logic to server now, as I may need to delete the GM from all existing categories, then move it to the default "channels" category of destination team. So now webapp won't be responsible for setting the category.

Reply:
Following the conversation on MM (https://community.mattermost.com/core/pl/16asd8t7z7dndfeoyi5869mjoh) I've removed the category setting logic from webapp and am setting the converted channel to the default "Channels" category for all GM members.